### PR TITLE
Sampler rework: refactor CWMH sampler

### LINKED ISSUE #371: Subclass the CWMH samplers from the new Sampler class

### PR DESCRIPTION
**Pull request by amal-ghamdi**
closes #371 

Things to consider for the review:
- The code is mainly copied from the original `CWMH`
- Modification to remove all the outer sampling loops and just keep the step and tune parts
- I needed to introduce additional scale parameter to keep track of old scale (to enable matching the old and new sampler)
- I modified the class documentation a bit, including stating that scale can be an array in this CWMH case. And updated the example code in the class documentation
-  I modified the tests method `assert_true_if_sampling_is_equivalent` and `assert_true_if_warmup_is_equivalent` to accept additional inputs: `old_idx` and `new_idx`. Which I use for example to remove first sample in the new sampler which is the initial point, that is not included in the samples of the old sampler. I use it also to remove the last sample in the old sampler because it is repeated (same as second to last sample)

Question:
- [x] I am not sure about the dim property? it seems like it was remover from the `NewSampler`. Should we still support it?(I removed dim from CWMH as discussed)

### CODE REVIEW

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:34]: I would suggest we remove the dim property. Instead user can define UserDefinedDistribution. We can discuss at next meeting.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:84]: We may want to include this in scale setter method for robustness?

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:92]: Consider adding a `#TODO` here related to issue 378. This overwriting of the proposal setter is a design we may reconsider.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:127]: I believe we can push this generic docstring to the base class. Other classes will inherit the docstring automatically.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:171]: Same comment about putting the generic docstring to base class

File: tests/experimental/test_mcmc.py
Comment (nabriis) [inline tests/experimental/test_mcmc.py:141]: Instead of old_idx and new_idx, which implicitly defines a range of values. I suggest for readabiltiy that we define an explicit range that we compare for the two samplers. E.g. one could compare [1..N] with [0..N-1] or similar. I think it would make the test much easier to understand.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:168]: In a way I like that you copied the sampler "step" explicitly. But since we are comparing with regression tests, I think it would be worthwhile to "clean up" some of the comments and improve the overall code quality of the step method in this new module.

Same comment holds for the warmup method.

**Review by nabriis — CHANGES_REQUESTED** [review]
Hi @amal-ghamdi. Thanks for adding the CWMH sampler in the module. The suggested changes look good overall, I simply had some minor comments in relation to the testing framework. In addition, I also had a comment on cleaning up the step and tune methods even more compared to the Sampler module.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py]: good idea! done

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py]: done

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py]: yes, that is great. I removed the doc string in step, tune, set_state and get_state.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py]: Nice, I updated the code for better readability, style and documentation

File: tests/experimental/test_mcmc.py
Comment (amal-ghamdi) [reply tests/experimental/test_mcmc.py:141]: I agree, this makes it clearer. I actually thought about this, but it can make it tricky to set default sample size and default range since the range depends on the sample size. Also setting Ns and range separately, adds overhead on the developer to make them consistent, i.e. sat for the same N.

**Comment by amal-ghamdi** [comment]
Thank you @nabriis for your helpful review! I addressed all your comments except the one about the dimension. We can talk about it today. I also added a comment about your suggested test change: from index to range.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [reply cuqi/experimental/mcmc/_cwmh.py:34]: After our discussion, we agreed to remove the "dim" property yes.

File: tests/experimental/test_mcmc.py
Comment (nabriis) [reply tests/experimental/test_mcmc.py:141]: Right. I think I got a bit confused. Perhaps this approach of using the indices for each set of samples makes sense. It may be helpful to add docstring with the parameters of this test.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [inline cuqi/experimental/mcmc/_cwmh.py:153]: This may be out of scope for this PR, but I wonder if we should get rid of the callable proprosal also? And just use distributions. Could create issue for this.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (nabriis) [reply cuqi/experimental/mcmc/_cwmh.py]: Thanks! It looks much better now.

**Review by nabriis — APPROVED** [review]
Thanks @amal-ghamdi. Looks good to me! I only had some small comments. I can approve it from here.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py:153]: good point, I added it to #373

File: tests/experimental/test_mcmc.py
Comment (amal-ghamdi) [reply tests/experimental/test_mcmc.py:167]: Great, I added documentation for these parameters

**Comment by amal-ghamdi** [comment]
Thank you @nabriis for your second review! I updated the PR to reflect your suggestions.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (jakobsj) [inline cuqi/experimental/mcmc/_cwmh.py:39]: Is the way callbacks are illustrated in demo31 still relevant for the new samplers? Maybe add an issue to update the demo once the samplers have replaced the old samplers.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (jakobsj) [inline cuqi/experimental/mcmc/_cwmh.py:79]: Is `dim` still available?  I understood from the PR comments it had been removed?

File: cuqi/experimental/mcmc/_cwmh.py
Comment (jakobsj) [inline cuqi/experimental/mcmc/_cwmh.py:102]: Do we have an issue to remember this? If not, add one?

**Review by jakobsj — APPROVED** [review]
Thanks @amal-ghamdi , from a quick look through I think it looks great and only have a couple of tiny questions added as individual comments, but approving now.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py:39]: good point, I recorded this in #378

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py:79]: `dim` as parameter to the initializer is removed. But we still have property `dim` extracted from the target dim.

File: cuqi/experimental/mcmc/_cwmh.py
Comment (amal-ghamdi) [reply cuqi/experimental/mcmc/_cwmh.py:102]: Thank you for pointing this out, I recorded it in #373

**Comment by amal-ghamdi** [comment]
@jakobsj , Many thanks for your review! I addressed your comments, feel free to resolve the conversations if you think the comments are addressed properly.

**Pull request merged**
